### PR TITLE
Detect recursive inline bindings and emit FS3888 diagnostic

### LINKED ISSUE #17991: Recursive use of inline member should produce better diagnostic
The following code
```fsharp

type ID3(vl: _) =
    member _.Value = vl

    member inline this.FastEquals(other: ID3) =
        this.FastEquals(other)
```

Is currently producing the following diagnostics:

```fsharp
    error FS1114: The value '_.ID3.FastEquals' was marked inline but was not bound in the optimization environment
    error FS1113: The value 'FastEquals' was marked inline but its implementation makes use of an internal or private function which is not sufficiently accessible
    error FS1114: The value '_.ID3.FastEquals' was marked inline but was not bound in the optimization environment
    error FS1113: The value 'FastEquals' was marked inline but its implementation makes use of an internal or private function which is not sufficiently accessible
    error FS1114: The value '_.ID3.FastEquals' was marked inline but was not bound in the optimization environment
    error FS1113: The value 'FastEquals' was marked inline but its implementation makes use of an internal or private function which is not sufficiently accessible
    error FS1114: The value '_.ID3.FastEquals' was marked inline but was not bound in the optimization environment
    error FS1113: The value 'FastEquals' was marked inline but its implementation makes use of an internal or private function which is not sufficiently accessible
```

which are quite misleading.

I propose we improve diagnostic with something like "Functions and members cannot be both inline and recursive".

### PR DESCRIPTION
**Pull request by T-Gro**
Fixes #17991

## Summary

Recursive use of an `inline` function or member now produces a single clear error (FS3890) instead of the misleading FS1113/FS1114/FS1116/FS1118 cascade from the optimizer.

A value or member marked `inline` is rejected when it takes part in a cycle made up entirely of `inline` bindings — direct self-recursion, or mutual recursion where every binding on the cycle is `inline`. A cycle routed through a non-`inline` sibling is left alone: inlining terminates at the non-`inline` binding, so such code keeps compiling as before.

Example now reported clearly:

```fsharp
type ID3(vl: int) =
    member _.Value = vl
    member inline this.FastEquals(other: ID3) = this.FastEquals(other)
```

### CODE REVIEW

**Review by T-Gro — COMMENTED** [review]
## Review: Detect recursive inline bindings and emit FS3888 diagnostic

Overall this is a well-designed fix that replaces the confusing cascade of FS1113/FS1114/FS1116/FS1118 with a single clear diagnostic. The error-recovery pattern (emit FS3888, then downgrade to `ValInline.Never`) is sound and prevents downstream cascading.

### Correctness

The cycle-detection algorithm is correct. One minor nit: the comment says *"perform BFS"* but the implementation uses `fv :: queue` (prepend) with `List.head`/`List.tail` (pop front), which is actually **DFS** (stack behavior). The result is still correct for cycle detection; only the comment is inaccurate.

### Tests

The new tests use raw diagnostic access (`r.Output.Diagnostics`) with `Assert.Equal`/`Assert.Contains` instead of the established component test DSL (`shouldFail |> withDiagnostics [...]`). Per the project's component-test instructions, the pipeline style is preferred. Consider rewriting, e.g.:

```fsharp
FSharp "module M\nlet rec inline f x = f x"
|> typecheck
|> shouldFail
|> withDiagnostics [
    (Error 3888, Line 2, Col 16, Line 2, Col 17,
     "The value or member 'f' has been marked 'inline' but is part of a recursive binding group...")
]
```

Also: the `Recursive inline member emits single clear diagnostic` test asserts exactly 1 diagnostic. If in the future additional warnings are emitted for the same code, this will break. Consider `withErrorCode 3888` or filtering by error code for robustness.

### Minor suggestions

1. **CheckDeclarations.fs** – the `ResizeArray` + `List.ofSeq` pattern is fine, but you could avoid the allocation by using a plain F# list accumulator reversed at the end, keeping the style more consistent with the surrounding functional code.

2. **Performance** – `freeInExpr CollectLocalsNoCaching e` is called per inline binding. Fine in practice (recursive inline groups are small), but worth a brief comment noting the assumption.

3. **typecheck vs compile** – the test `Non-recursive inline in rec group compiles cleanly` and `Inline CE builder members do not trigger recursive diagnostic` only need type-checking. Using `typecheck` instead of `compile` would be faster on CI per the test guidelines.

### Summary

:white_check_mark: Algorithmic correctness looks good
:white_check_mark: Good error recovery via `SetInlineInfo`
:white_check_mark: Covers both module-level and class-member paths
:warning: Test style diverges from DSL conventions
:information_source: Comment says BFS but code is DFS (correct either way for this purpose)

File: src/Compiler/Checking/Expressions/CheckExpressions.fs [hunk 1]
```diff
@@ -13071,6 +13071,52 @@ and FixupLetrecBind (cenv: cenv) denv generalizedTyparsForRecursiveBlock (bind:
...
+            let mutable queue = [startStamp]
+            while not (List.isEmpty queue) && not foundCycle do
+                let cur = List.head queue
+                queue <- List.tail queue
+                let frees = freeStampsByStamp |> Map.tryFind cur |> Option.defaultValue Set.empty
+                for fv in frees do
+                    if not foundCycle then
+                        if fv = startStamp then
+                            foundCycle <- true
+                        elif Set.contains fv inlineStamps && not (Set.contains fv visited) then
```
Comment (T-Gro) [inline src/Compiler/Checking/Expressions/CheckExpressions.fs]: **[Correctness — false negative]** The BFS only enqueues callees that are themselves in `inlineStamps` (`Set.contains fv inlineStamps && ...`). That misses any cycle that routes through a *non-inline* sibling in the same letrec group, e.g.:

```fsharp
let rec inline f x = g x
and g x = f x
```

Here `f` is genuinely recursive (cycle `f → g → f`), but `g ∉ inlineStamps`, so BFS from `f` discards the only outgoing edge and reports no cycle. The user then gets the original FS1113/FS1114/FS1116/FS1118 cascade from the optimizer — exactly the regression #17991 was filed to prevent.

Fix: build `freeStampsByStamp` for **all** bindings in the group (not just inline ones) and drop the `Set.contains fv inlineStamps` guard in the BFS. The `fv = startStamp` check is the only correct cycle predicate.

File: tests/FSharp.Compiler.ComponentTests/Conformance/InferenceProcedures/RecursiveSafetyAnalysis/RecursiveSafetyAnalysis.fs [hunk 2]
```diff
@@ -117,3 +115,60 @@ module RecursiveSafetyAnalysis =
...
+            |> fun r -> r.Output.Diagnostics
+        Assert.Equal(1, diags.Length)
+        Assert.Contains("recursive", diags.[0].Message, StringComparison.OrdinalIgnoreCase)
+
+    [<Fact>]
+    let ``Non-recursive inline in rec group compiles cleanly`` () =
+        let source = "module M\nlet rec inline f x = x + 1"
+        FSharp source
+        |> compile
+        |> shouldSucceed
```
Comment (T-Gro) [inline tests/FSharp.Compiler.ComponentTests/Conformance/InferenceProcedures/RecursiveSafetyAnalysis/RecursiveSafetyAnalysis.fs:158]: **[Test Coverage]** Missing the most informative case — an `inline` binding whose cycle passes through a *non-inline* sibling:

```fsharp
let rec inline f x = g x
and g x = f x
```

Adding this as a positive test (expect FS3888 on `f`) would have caught the inline-only-edge bug in `CheckRecursiveInlineGroup`. Please also add the matching negative test:

```fsharp
let rec inline f x = g x
and g x = x + 1   // no cycle back to f -> no diagnostic
```

Without these two, future refactors can silently regress the detector to direct-self-reference-only semantics.

File: src/Compiler/Checking/Expressions/CheckExpressions.fs [hunk 3]
```diff
@@ -13071,6 +13071,52 @@ and FixupLetrecBind (cenv: cenv) denv generalizedTyparsForRecursiveBlock (bind:
...
+                            visited <- Set.add fv visited
+                            queue <- fv :: queue
+            if foundCycle then
+                errorR(Error(FSComp.SR.tcRecursiveInlineNotAllowed(v.DisplayName), v.Range))
+                v.SetInlineInfo ValInline.Never
```
Comment (T-Gro) [inline src/Compiler/Checking/Expressions/CheckExpressions.fs:13219]: **[FCS API Surface]** `v.SetInlineInfo ValInline.Never` mutates the `Val` flags after `errorR`. In the language service the partially-checked symbol table is exposed to IDE clients via `FSharpMemberOrFunctionOrValue.InlineAnnotation`; once this branch fires, tooltips/analyzers will report the binding as non-inline even though the user wrote `inline`. The project won't compile (hard error), but the symbol is still surfaced to IDE consumers during editing. Worth either (a) adding a code comment documenting this trade-off, or (b) leaving the flag alone and letting the optimizer skip recursive-inline'd vals when prior errors exist (closer to a conservative fix).

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 4]
```diff
@@ -1501,52 +1501,61 @@ module MutRecBindingChecking =
...
+                            collectedBinds.Add pgbrind
+                            yield pgbrind ])
 
-            // Phase2C: Fixup let bindings 
-            (fun bindIdxs -> 
-                    [ for idx in bindIdxs do 
-                        let generalizedBinding = generalizedBindingsMap[idx] 
-                        let vxbind = TcLetrecAdjustMemberForSpecialVals cenv generalizedBinding
-                        yield FixupLetrecBind cenv denv generalizedTyparsForRecursiveBlock vxbind ])
+        CheckRecursiveInlineGroup (List.ofSeq collectedBinds)
```
Comment (T-Gro) [inline src/Compiler/Checking/CheckDeclarations.fs:1557]: **[Scope / Double-Check]** `collectedBinds` aggregates every fixed-up binding across *all* tycons and *all* let-groups in the MutRec block before invoking `CheckRecursiveInlineGroup`. That is fine for cross-type mutual recursion, but please verify there's no double-diagnosis when an inline `let rec` appears inside a member body (handled both here and from `TcLetrecBindings`'s own call at line 13153). A quick test with a nested `let rec inline` inside a member would confirm.

File: src/Compiler/Checking/Expressions/CheckExpressions.fs [hunk 5]
```diff
@@ -13071,6 +13071,52 @@ and FixupLetrecBind (cenv: cenv) denv generalizedTyparsForRecursiveBlock (bind:
...
+        // through other inline bindings counts.
+        for pgrbind in inlineBindings do
+            let (TBind(v, _, _)) = pgrbind.Binding
+            let startStamp = v.Stamp
+            let mutable foundCycle = false
+            let mutable visited = Set.empty
+            let mutable queue = [startStamp]
+            while not (List.isEmpty queue) && not foundCycle do
+                let cur = List.head queue
+                queue <- List.tail queue
```
Comment (T-Gro) [inline src/Compiler/Checking/Expressions/CheckExpressions.fs]: **[Code Style — minor]** Using a `list` as a worklist with `List.head` / `queue <- List.tail queue` is O(n) per pop and allocates a fresh tail each time. For consistency with how the compiler typically writes worklists, prefer `System.Collections.Generic.Queue<_>` or a mutable `Stack<_>` (order doesn't matter for reachability). Negligible for realistic group sizes but cleaner.

**Review by T-Gro — COMMENTED** [review]
🤖 AI-generated review (expert-reviewer agent). Flagging actionable findings only.

File: src/Compiler/Checking/Expressions/CheckExpressions.fs [hunk 6]
```diff
@@ -13071,6 +13071,52 @@ and FixupLetrecBind (cenv: cenv) denv generalizedTyparsForRecursiveBlock (bind:
...
+        // Map from inline stamp to set of free-local stamps in its body.
+        let freeStampsByStamp =
+            inlineBindings
+            |> List.map (fun pgrbind ->
+                let (TBind(v, e, _)) = pgrbind.Binding
+                let freeVals = (freeInExpr CollectLocalsNoCaching e).FreeLocals
+                let frees = Zset.fold (fun (fv: Val) acc -> Set.add fv.Stamp acc) freeVals Set.empty
+                v.Stamp, frees)
+            |> Map.ofList
+        // For each inline binding, perform BFS through inline-only edges and
```
Comment (abonie) [inline src/Compiler/Checking/Expressions/CheckExpressions.fs]: This seems to be DFS not BFS
A list used like this
```fsharp
let cur = List.head queue
queue <- List.tail queue
```
and
```fsharp
queue <- fv :: queue
```
is a LIFO stack, not a FIFO queue. I don't think this distinction matters here? But the comment is misleading

File: tests/FSharp.Compiler.ComponentTests/Conformance/InferenceProcedures/RecursiveSafetyAnalysis/RecursiveSafetyAnalysis.fs [hunk 7]
```diff
@@ -117,3 +115,60 @@ module RecursiveSafetyAnalysis =
...
+    member inline _.X = self.Y
+    member inline _.Y = self.X"""
+        let diags =
+            FSharp source
+            |> compile
+            |> fun r -> r.Output.Diagnostics
+        let recursiveErrs =
+            diags
+            |> List.filter (fun d -> d.Message.IndexOf("recursive", StringComparison.OrdinalIgnoreCase) >= 0)
+        Assert.True(recursiveErrs.Length >= 1)
```
Comment (abonie) [inline tests/FSharp.Compiler.ComponentTests/Conformance/InferenceProcedures/RecursiveSafetyAnalysis/RecursiveSafetyAnalysis.fs]: Why not test `= 2` instead of `>= 1` here?

**Review by abonie — APPROVED** [review]

File: src/Compiler/Checking/Expressions/CheckExpressions.fs (hunk 1 above)
Comment (T-Gro) [reply src/Compiler/Checking/Expressions/CheckExpressions.fs]: Keeping the guard. `let rec inline f x = g x and g x = f x` compiles today — the cycle runs through non-inline `g`, so inlining terminates there. Only all-inline cycles are rejected; broadening to every cycle would reject valid terminating code. Added pos/neg tests.

File: tests/FSharp.Compiler.ComponentTests/Conformance/InferenceProcedures/RecursiveSafetyAnalysis/RecursiveSafetyAnalysis.fs (hunk 2 above)
Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/Conformance/InferenceProcedures/RecursiveSafetyAnalysis/RecursiveSafetyAnalysis.fs:158]: Added a non-inline-sibling compile case plus a 3-hop all-inline cycle asserting FS3890.

File: src/Compiler/Checking/Expressions/CheckExpressions.fs (hunk 3 above)
Comment (T-Gro) [reply src/Compiler/Checking/Expressions/CheckExpressions.fs:13219]: Kept (a). Comment now documents the cascade suppression and the IDE trade-off (binding briefly surfaces as non-inline).

File: src/Compiler/Checking/CheckDeclarations.fs (hunk 4 above)
Comment (T-Gro) [reply src/Compiler/Checking/CheckDeclarations.fs:1557]: No duplicate diagnostic — a nested `let rec inline` inside a member reports once. Test added.

File: src/Compiler/Checking/Expressions/CheckExpressions.fs (hunk 5 above)
Comment (T-Gro) [reply src/Compiler/Checking/Expressions/CheckExpressions.fs]: Switched to a FIFO `Queue` + `HashSet`; comment corrected.

File: src/Compiler/Checking/Expressions/CheckExpressions.fs (hunk 6 above)
Comment (T-Gro) [reply src/Compiler/Checking/Expressions/CheckExpressions.fs]: Right — now a real FIFO `Queue`, comment fixed.

File: tests/FSharp.Compiler.ComponentTests/Conformance/InferenceProcedures/RecursiveSafetyAnalysis/RecursiveSafetyAnalysis.fs (hunk 7 above)
Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/Conformance/InferenceProcedures/RecursiveSafetyAnalysis/RecursiveSafetyAnalysis.fs]: Now asserts exactly two FS3890.

**Review by abonie — APPROVED** [review]
